crates/circuit/src/dag_circuit.rs

```diff
@@ -2919,43 +2919,69 @@ def _format(operand):
 
     /// Remove all of the ancestor operation nodes of node.
     fn remove_ancestors_of(&mut self, node: &DAGNode) -> PyResult<()> {
-        // anc = rx.ancestors(self._multi_graph, node)
-        // # TODO: probably better to do all at once using
-        // # multi_graph.remove_nodes_from; same for related functions ...
-        //
-        // for anc_node in anc:
-        //     if isinstance(anc_node, DAGOpNode):
-        //         self.remove_op_node(anc_node)
-        todo!()
+        let dag_binding = self.dag.clone();
+        for ancestor in core_ancestors(&dag_binding, node.node.unwrap())
+            .filter(|next| next != &node.node.unwrap())
+            .filter(|next| match dag_binding.node_weight(*next) {
+                Some(NodeType::Operation(_)) => true,
+                _ => false,
+            })
+        {
+            self.dag.remove_node(ancestor);
+        }
```

Comment:
I'm pretty sure cloning `DAGCircuit` is expensive (but I have just now looked at it)
This would probably be more performant, although I did not test it.
```suggestion
        let ancestors: Vec<_> = core_ancestors(&self.dag, node.node.unwrap())
            .filter(|next| {
                next != &node.node.unwrap()
                    && match self.dag.node_weight(*next) {
                        Some(NodeType::Operation(_)) => true,
                        _ => false,
                    }
            })
            .collect();
        ancestors.iter().map(|a| self.dag.remove_node(*a));
```
I also put the two conditions in a single filter function.

Reply:
It'd be good to do the approach of first collecting the node IDs into a `Vec` as John has done here instead of cloning the DAG, since the latter is much more expensive.

The filter thing is likely personal preference, since the compiler is almost certainly smart enough to remove the overhead.

For the `ancestors.iter().map(|a| self.dag.remove_node(*a))` in John's suggestion, you'd probably need to do:

```rust
for a in ancestors {
    self.dag.remove_node(a)
}
```

since I don't think the `map` by itself will actually pump the iterator.

Reply:
> The filter thing is likely personal preference,

Sure  the compiler will probably optimize this away, and even if it didn't you'd probably never notice.

Reply:
It's pretty common for people to require handling variants explicitly instead of `_ => false`. But anything I can think of doing here is so much less readable. Like this:
```rust
if let Some(node) = self.dag.node_weight(*next) {
                        match node {
                            NodeType::Operation(_) => true,
                            NodeType::ClbitIn(_) | NodeType::ClbitOut(_) | NodeType::QubitIn(_) | NodeType::QubitOut(_) => false,
                        }
                    }
                else {
                    false
                }
```